modin/pandas/groupby.py

```diff
@@ -171,10 +173,76 @@ def idxmax(self):
     def ndim(self):
         return 2  # ndim is always 2 for DataFrames
 
-    def shift(self, periods=1, freq=None, axis=0):
-        return self._default_to_pandas(
-            lambda df: df.shift(periods=periods, freq=freq, axis=axis)
-        )
+    def shift(self, periods=1, freq=None, axis=0, fill_value=None):
+        """
+        Shift each group by periods observations.
+
+        If freq is passed, the index will be increased using the periods and the freq.
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Number of periods to shift.
+        freq : str, optional
+            Frequency string.
+        axis : axis to shift, default 0
+            Shift direction.
+        fill_value : optional
+            The scalar value to use for newly introduced missing values.
+
+        Returns
+        -------
+        Series or DataFrame
+            Object shifted within each group.
+        """
+
+        def _shift(periods, freq, axis, fill_value, is_set_nan_rows=True):
+            from .dataframe import DataFrame
+
+            result = self._df.shift(periods, freq, axis, fill_value)
+
+            if is_set_nan_rows:
+                if (
+                    isinstance(self._by, BaseQueryCompiler)
+                    and set(self._by.columns).issubset(list(self._df.columns))
```

Comment:
```suggestion
                    and len(self._by.columns.unique.sort_values().difference(self._df.columns.unique().sort_values())) == 0
```
It seems that conventions `pandas.Index` to set/list are very expensive. I don't suggest the solution above, it's really greedy, maybe there is some more efficient way to determine is columns of by are subset of frame columns
```
sort 2.0272323302924633
set 3.8749265801161528
```


<details><summary> Script to measure </summary>


```python
import pandas
import numpy

N = 10_000_000
ind1 = pandas.Index(np.arange(N))
ind2 = pandas.Index(np.arange(N))

t1 = timer()
ind1_s = ind1.unique().sort_values()
ind2_s = ind2.unique().sort_values()
diff = len(ind2_s.difference(ind1_s))
print("sort", timer() - t1)

t1 = timer()
res = set(ind1).issubset(list(ind2))
print("set", timer() - t1)

```


</details>

(please verify this results with different types of index)

Reply:
I tested this with `MultiIndex` and `DateTimeIndex`. Results:

DateTimeIndex
```
sort 2.0223687179386616
set 41.05154676735401
```
MultiIndex
```
sort 102.30562068708241      
set 12.652700988575816 
```
Index
```
sort 2.167663438245654
set 4.2023481242358685
```

Based on result I'll use `set` in only if `self._by.columns` is `MultiIndex`
<details><summary>script</summary>

```python
import pandas
import numpy as np
from timeit import default_timer as timer


N = 10_000_000

# Index
ind1 = pandas.Index(np.arange(N))
ind2 = pandas.Index(np.arange(N))

# MultiIndex
# ind1 = pandas.MultiIndex.from_arrays([np.arange(N), np.arange(N)])
# ind2 = pandas.MultiIndex.from_arrays([np.arange(N), np.arange(N)])

# DateTimeIndex
# ind1 = pandas.date_range('12/12/1999', periods=N, freq='NS')
# ind2 = pandas.date_range('12/12/1999', periods=N, freq='NS')

t1 = timer()
ind1_s = ind1.unique().sort_values()
ind2_s = ind2.unique().sort_values()
diff = len(ind2_s.difference(ind1_s))
print("sort", timer() - t1)

t1 = timer()
res = set(ind1).issubset(list(ind2))
print("set", timer() - t1)
```
</details>

Reply:
`set` is used only if `self._by.columns` is `MultiIndex`

Reply:
I've also found that the index subset could be determined with [`pandas.Index.isin()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.isin.html) method

could you please also measure this approach?
```python
by.columns.isin(df.columns).all()
```

Reply:
Results are presented below.

Index
```
sort 2.159685555845499
set 4.1990830190479755
isin 3.2081973012536764
```

MultiIndex
```
sort 101.13373468071222
set 12.397226087749004
isin 31.542536107823253
```

DateTimeIndex
```
sort 2.022602867335081
set 40.36840843223035
isin 1.1622390747070312
```

Looks like `isin` is the best way when `index` is `DateTimeIndex`